iptables: Keep old rules while adding new ones

jrajahalme opened the pull request:
Keep old iptables rules by renaming Cilium chains so that new rules
can be added while old are still in use. Copy old TPROXY rules from
the renamed old rules. Remove the backups only after new rules have
been successfully added, but keep the copied TPROXY rules, as the
new ones may be installed considerably later.

On first initialization only copy over the DNS proxy TPROXY rules, as
other proxies can't reuse old proxy ports across restarts.

Pick the last applicable proxy port from iptables, if multiple are
present.

Remove stale TPROXY rules once the current port is known and allow
for the case where the TPROXY rule already exists (e.g., from the copy
done from old rules).

This change makes it possible to keep old rules in effect while adding
new ones without special consideration for transient rules.

Fixes: #16364 

```release-note
DNS proxy is now more available during Cilium restarts, including upgrades.
```

jrajahalme committed:
iptables: Keep old rules while adding new ones

Keep old iptables rules by renaming Cilium chains so that new rules
can be added while old are still in use. Copy old TPROXY rules from
the renamed old rules. Remove the backups only after new rules have
been successfully added.

This change makes it possible to keep old rules in effect while adding
new ones without special consideration for transient rules.

On first initialization only copy over the DNS proxy TPROXY rules, as
other proxies can't reuse old proxy ports across restarts.

Pick the last applicable proxy port from iptables, if multiple are
present.

Remove stale TPROXY rules once the current port is known.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
endpoint: Do not panic in Finalize()

Panicing in Finalize functions may leave endpoint locked and brick the
whole agent. Better avoid itt and log errors instead, and unlock the
Endpoint in defer if it still happens.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

ti-mo commented on pkg/endpoint/bpf.go:
```diff
@@ -955,9 +955,9 @@ func (e *Endpoint) finalizeProxyState(regenContext *regenerationContext, err err
+		defer e.unlock() // In case Finalize() panics
 		e.getLogger().Debug("Finalizing successful endpoint regeneration")
 		datapathRegenCtx.finalizeList.Finalize()
-		e.unlock()
```
Doesn't e`.unlock()` following `Revert()` below have the same issue?

jrajahalme replied:
Good catch - it does.

jrajahalme replied:
Fixed.

joestringer commented on pkg/datapath/loader/base.go:
```diff
@@ -453,25 +453,11 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
-		if err != nil {
-			return err
-		}
+	if err := iptMgr.InstallRules(option.Config.HostDevice, firstInitialization, option.Config.InstallIptRules); err != nil {
```
I'm confused, I was under the impression that the "Transient rules" logic here was specifically to address a similar type of problem that this PR is proposing to solve (commit 80507f142c59a5d0ca6b782e6b8092e0c93a7246) though maybe with some other nuance like if iptables is configured with a default drop then we must retain some sort of allow rule in place while we figure out the real full ruleset that will be needed.

I don't see any consideration in the commit message for the problem that this 'transient rules' logic was trying to solve or rationale for why we don't need it any more. Are you able to describe that?

jrajahalme replied:
"Transient rules" logic partially solves the problem of traffic getting dropped between removal of Cilium chains and adding them back up again. "Partially" because the transient rules only cover a part of the rule set (no consideration for TPROXY rules, for example).

In essence, this PR keeps all old rules in place as "transient rules" by renaming all Cilium chains instead of removing them. These renamed chains are removed only after the new rules have been added to the new Cilium chains.

Commit message states:
```
This change makes it possible to keep old rules in effect while adding
new ones without special consideration for transient rules.
```
I'll try to elaborate that a bit.

joestringer replied:
👍 OK that motivation makes sense to me. This also explains why the atomicity of iptables chain rename is such an important question to answer.

joestringer commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -840,9 +895,13 @@ func (m *IptablesManager) GetProxyPort(name string) uint16 {
-	str := re.FindString(string(res))
-	portStr := re.ReplaceAllString(str, "$1")
+	re := regexp.MustCompile(name + ".*TPROXY redirect (0.0.0.0|::):([1-9][0-9]*) mark")
+	strs := re.FindAllString(string(res), -1)
+	if len(strs) == 0 {
+		return 0
+	}
```
What's the impact or need for this particular change in this commit, in context of the rest of the changes proposed here? Is it possible to put this in a separate commit and describe why this is necessary, or is it integral to the overall approach?

jrajahalme replied:
Previously only IPv4 address format (`0.0.0.0`) was detected, with this change also IPv6 address format (`::`) is supported. This is a separate issue, but the rest here is integral to the new approach:

Instead of finding the first match and getting the proxy port from there we need to find the last (== latest) match and get the proxy port from there. This is significant if/when there are multiple TPROXY rules for the same proxy (e.g., DNS), some old and the last one new. This could happen if on restart an explicit proxy port is configured and we failed to clean up the old rule from the chain. Without this change the next restart would then try to use the stale port of the 1st match. We do clean up the old rules, but if Cilium gets killed right before the cleanup we may find multiple rules in the chain for the same proxy.

New rules are added with `-A`, so they are always last in the chain.

joestringer replied:
This is quite nuanced, do you think it should be encoded as a comment or in the git logs as well?

jrajahalme replied:
The new unit tests cover this too :-)

joestringer commented on pkg/proxy/proxy.go:
```diff
@@ -262,22 +261,17 @@ func (p *Proxy) ackProxyPort(pp *ProxyPort) error {
-		// Remove old rules, if any and for different port
-		if pp.rulesPort != 0 && pp.rulesPort != pp.proxyPort {
-			scopedLog.Infof("Removing old proxy port rules for %s:%d", pp.name, pp.rulesPort)
-			p.datapathUpdater.RemoveProxyRules(pp.rulesPort, pp.ingress, pp.name)
```
May be a naive question, but why was this required before and what logic replaces this? Could there be some stale state issue because this is no longer run here, or is it simply moved elsewhere?

(if at all possible, splitting this particular change into another commit and explaining it would help a lot for understanding the implications of the change).

jrajahalme replied:
Removal of stale TPROXY rules is now done in the iptables package when new TPROXY rules are added. It is needed when the port bound to the proxy is different from the old one. Normally we reuse the old port, but we do support passing the DNS proxy port as a command line option, I recall.

joestringer commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -231,21 +226,32 @@ func (m *IptablesManager) removeCiliumRules(table, prog, match string) {
-				log.Warnf(string(combinedOutput))
-				log.WithError(err).WithField(logfields.Object, args).Warnf("Unable to process chain %s with %s (%s)", c.name, prog, operation)
-			}
+		if err != nil && !quiet {
```
Incidentally I noticed a weird trend in this code where we pass "quiet=true" as the last parameter to `runProgCombinedOutput()` then check the error and `quiet` to decide whether to log or not. 

This makes sense if we have to check the error to decide to return early from the function, but not just to log the error. We don't need to do this error check / quiet bit, just pass `quiet` down into `runProgCombinedOutput()`. It'll even do the correct structured logging instead of using `Warnf()`.

joestringer replied:
^^ It occurred to me that maybe this kind of code is trying to work around some discrepancy in the underlying CombinedOutput() implementation and how it handles logging, but if that's the case it'd be nice to properly address that in the common code instead of wrapping it all over the place to work around the weird behaviour.

jrajahalme replied:
You are likely right, but not really understanding all this quiet stuff I tried not to touch that part of the logic here. My understanding is that generally we want to log errors, but sometimes those are more confusing than helpful and we've decided not to log them to avoid the confusion.

IMO cleaning up the logging behavior is a separate task that has it's own risks and testing requirements and is better addressed in a separate PR, should we decide to invest in it.

joestringer replied:
Yeah you raise a good point here, the logging behaviour in that interface is weird and that's what's causing this particular pain point. I agree not necessary for this PR and could do with some dedicated attention.

joestringer commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -695,26 +677,99 @@ func (m *IptablesManager) installStaticProxyRules() error {
+
+// copies old proxy rules
+func (m *IptablesManager) copyProxyRules(oldChain string, match string) {
+	copyRules := func(prog string, table string, re *regexp.Regexp, match, oldChain, newChain string) {
```
I had a bit of trouble putting my finger on it, but these callback functions seemed a bit odd to me. I think the reason is that they are inherently untestable. Since it's a variable inside another function, we can't validate that it is performing the right actions. I realize that the outer function could theoretically be tested, but then we definitely bring in the iptables aspect to this, manage that kernel dependency during testing and so on. In an ideal world I feel like it'd be nice to have full-on unit tests that provide examples of what the iptables rules might look like (even if we had one giant static ruleset defined at the top), then we could run various functions like the scan/match piece of this function, and unit test to validate that this has the right behaviour.

I'm not going to push on this since this is a pretty targeted fix which is aimed for backporting, but this is clearly an aspect of technical debt we've picked up along the way. We're already having trouble encouraging people to get involved in reviewing code like this. Sure, part of that is just the nature of iptables and the difficulty understanding  the repercussions. But there's also just large testing gaps so even if someone wanted to pick up this code and play around with it, it's difficult to reason about or validate even the simplest of operations.

jrajahalme replied:
introduced this local function to avoid repeating code; technically it is not a callback function as we do not return it. Will make this a separate package level function and add unit testing with some way of mocking up the `prog`.

joestringer commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -695,26 +677,99 @@ func (m *IptablesManager) installStaticProxyRules() error {
+	copyRules := func(prog string, table string, re *regexp.Regexp, match, oldChain, newChain string) {
+		output, err := runProgCombinedOutput(prog, append(m.waitArgs, "-t", table, "-S"), true)
+		if err != nil {
+			log.WithError(err).Infof("Can't list %s %s table rules", prog, table)
```
Can we make this warning clearer to the user? There are no consequences in the current message, just a random info message that reads more like a debug message for developers.

Suggestion:

```
log.WithFields(logrus.Fields{
    "table": table,
    "prog": prog,
}).WithError(err).Warning("Cannot list rules in table. May reallocate layer 7 proxy port which could cause disruption for traffic selected by L7 policy", prog, table)
```

Then if there's a false positive in some common cases (like first startup) then maybe we can filter on the err output to skip the log in the known case like where the table is not available because we didn't create it yet?

jrajahalme replied:
Fixed, thanks!

joestringer commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -695,26 +677,99 @@ func (m *IptablesManager) installStaticProxyRules() error {
+				log.WithField(logfields.Object, logfields.Repr(rule)).Debugf("Considering copying %s TPROXY rule from %s to %s", prog, oldChain, newChain)
+				args, err := shellwords.Parse(strings.Replace(rule, oldChain, newChain, 1))
+				if err != nil {
+					log.WithError(err).WithField(logfields.Object, rule).Warnf("Unable to parse %s TPROXY rule", prog)
```
Ditto here, I think we should log the consequence (potential impact on traffic selected by L7 rules). Presumably same below too.

jrajahalme replied:
This is much less probable to surface to the logs (iptables format would need to change).
Fixed.

joestringer commented on pkg/proxy/proxy.go:
```diff
@@ -520,10 +519,7 @@ func (p *Proxy) CreateOrUpdateRedirect(l4 policy.ProxyPolicy, id string, localEn
-					// happen if there is an internal Cilium logic error regarding
-					// installation of iptables rules.
-					panic(err)
+					log.WithError(err).Errorf("ackProxyPort failed on %s", pp.name)
```
What does `ackProxyPort failed` mean to the user? Can we make this less terse to give the user better clues if this happens, maybe some notion of the impact on policy implementation or something?

jrajahalme replied:
Changed this to "Datapath proxy redirection cannot be enabled for %s, L7 proxy may be bypassed". The error contains more detail, such as `fmt.Errorf("Cannot install proxy rules for %s: %s", pp.name, err)`

joestringer commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -1199,7 +1250,7 @@ func (m *IptablesManager) InstallRules(ifName string) error {
 
 	if err := m.installForwardChainRules(ifName, localDeliveryInterface, ciliumForwardChain); err != nil {
-		return fmt.Errorf("cannot install forward chain rules to %s: %s", transientChain.name, err)
+		return fmt.Errorf("cannot install forward chain rules to %s: %s", ciliumForwardChain, err)
```
nit: We should really wrap the error here rather than formatting, `%w` at the end of the string.

jrajahalme replied:
Fixed.

joestringer reviewed (approved):
The logic seems right, my main concern is around giving the user friendly errors which they could be more likely to understand / reason about and potentially take action.

I didn't pay super close attention to the core string scanning / matching logic but really we should just have unit tests for that with sample inputs to prove it does the right thing. I don't really see how someone can review the logic without sitting side-by-side with the expected inputs and reason about what's going on. One of the threads below goes into a bit more detail.

There's a few other nits I picked on below, though I tried to keep to a minimum - general code style stuff we've picked up along the way like properly wrapping errors and using structured logs. On one hand I feel that we're trying not to spend too much time cleaning up this iptables-handling code when the roadmap involves deprecating it entirely, but at the same time I feel that it's painful each and every time we touch it. :/

jrajahalme committed:
iptables: Add rudimentary unit testing

Wrap "iptables" and "ip6tables" programs with iptablesInterface so
that unit testing can mock up the executables.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

joestringer commented on pkg/datapath/iptables/iptables_test.go:
```diff
@@ -0,0 +1,510 @@
+	}
+
+	// New port number, adds new ones, deletes stale rules. Does not touch OLD_ chains
+	mockManager.addProxyRules(mockIp4tables, 43479, false, "cilium-dns-egress")
```
If I read this right, we end up with 2 add commands and 2 delete commands for the `CILIUM_PRE_mangle` chain, but `CILIUM_PRE_mangle` doesn't have any existing rules. The comment on line 290 implies that we shouldn't be touching `OLD_CILIUM_PRE_mangle` chain, so that is satisfied... but why are the deletes being issued? Seems like they are no-ops if there is no such existing rule?

(Ditto for IPv6)

jrajahalme replied:
Thanks for noticing this :-) Those delete expectations were an accidental copy&paste remnant and the test code did not check for unmet expectations at the end of each test. I added that checking and removed these delete expectations.

joestringer commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -689,52 +687,135 @@ func (m *IptablesManager) installStaticProxyRules() error {
+				log.WithFields(logrus.Fields{
+					"table": table,
+					"prog":  prog.getProg(),
+				}).WithError(err).WithField(logfields.Object, rule).Warnf("Unable to parse %s TPROXY rule, disruption to traffic seleccted by L7 policy possible", prog)
```
nit, if `prog` is part of the structured logging, then I don't think it needs to be part of the message any more. In general I encourage avoiding the Warnf() variations on logging functions but in practice if the variable content is only a couple of variations it's not as bad as if it can be any random dynamic content. We also can combine the `WithFields()` and `WithField()`. There's also a typo here.

```suggestion
				log.WithFields(logrus.Fields{
					"table": table,
					"prog":  prog.getProg(),
					logfields.Object: rule,
				}).WithError(err).Warn("Unable to parse TPROXY rule, disruption to traffic selected by L7 policy possible")
```

(Similar feedback for other log statements in the PR).

jrajahalme replied:
I'm convinced by now that my keyboard is acting up, doubling letters randomly...
Fixed :-)

joestringer replied:
Not gonna point my fingers at any particular fruit-branded manufacturers... ;)

joestringer reviewed (changes requested):
Thanks, the new testing makes me much more confident. Couple more nits I noticed in the testing and in the log messages.

joestringer reviewed (approved):
Latest changes LGTM, thanks!
